File: RemoveKDigit.java
// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach

import java.util.*;

public class RemoveKDigit {
    
        public String removeKdigits(String num, int k) {
    
            Stack<Character> st = new Stack<>();
            int i =0;
    
            while (i < num.length()){
                while(!st.isEmpty() && k>0 && st.peek() > num.charAt(i)){
                    st.pop();
                    k--;
                }
                st.push(num.charAt(i));
                i++;
            }
    
            while (k>0){
                st.pop();
                k--;
            }
    
            StringBuilder sb = new StringBuilder();
            while (!st.isEmpty()){
                sb.append(st.pop());
            }
    
            //String result = "";
            //boolean zero = true;
            sb.reverse();
            System.out.println(sb.toString());
            while (sb.length() > 1 && sb.charAt(0) == '0'){
                sb.deleteCharAt(0);
            }
    
            //return sb.reverse().toString();
            if (sb.toString().equals("")) return "0";
            return sb.toString();
    
        }
    }

File: RemoveKDigit.py
class Solution:
    def removeKdigits(self, num: str, k: int) -> str:
        if(k==0):
            return num
        if(len(num)==1 and k==1):
            return "0"
        st = []
        o = []
        i=0
        while(i<len(num)):
            while(k>0 and len(st)>0 and int(num[i])<int(st[-1])):
                st.pop()
                k-=1
            st.append(num[i])
            i+=1
        
        st = st[:-k] if k>0 else st
        
        return "".join(st).lstrip('0') or "0"
